### PR TITLE
RemoveDuplicateTags func introduced to fix #2514

### LINKED ISSUE #2514: @goTag do not override existing tag
`@goTag` do not override existing tag

Ideally it should override existing tag. sole purpose of the directive is to give control to user.
Looks

image

image

`GoTagFieldHook` just appends new tags

https://github.com/99designs/gqlgen/blob/866187fd2510121d1b5f0d0636c8d37d80191c91/plugin/modelgen/models.go#L415

### PR DESCRIPTION
**Pull request by valllabh**
As described in #2514, and without this PR, `GoTagFieldHook` just appends new tags, so `@goTag` will not override an existing tag.

Ideally it should override existing tag, since the sole purpose of the `@goTag` directive is to give control to user.

In this PR, the `removeDuplicateTags` helper func is introduced to fix #2514

I have:
 - [✓] Added tests covering the bug / feature (see [testing](https://github.com/99designs/gqlgen/blob/master/TESTING.md))
 - [Not Required] Updated any relevant documentation (see [docs](https://github.com/99designs/gqlgen/tree/master/docs/content))

### CODE REVIEW

**Comment by StevenACoffman** [comment]
Thanks for working on this!

The first run through only some of the matrix of tests failed (e.g. ubuntu, Go 1.18):
```
--- FAIL: TestModelGeneration (18.60s)
    --- FAIL: TestModelGeneration/field_hooks_are_applied (0.00s)
        models_test.go:92: 
            	Error Trace:	models_test.go:92
            	Error:      	Should be true
            	Test:       	TestModelGeneration/field_hooks_are_applied
FAIL
FAIL	github.com/99designs/gqlgen/plugin/modelgen	37.009s
```

I wonder if this means that either in that particular test, or perhaps in your code, it somehow depends on an order that is not deterministic. I'm on my phone right now, so I can't look at it in more detail, but just food for thought.

**Comment by StevenACoffman** [comment]
Also, your code appears to prefer the last value of duplicate tags, rather than the first. Is there a reason for this?

I flipped it the other way in this example which allows the code to be simplified:
https://go.dev/play/p/yQfwGrFfOGo

```
// removeDuplicateTags removes duplicate tags
func removeDuplicateTags(t string) string {

	processed := make(map[string]struct{})

	tt := strings.Split(t, " ")
	returnTags := ""

	for _, ti := range tt {
		kv := strings.Split(ti, ":")
		if len(kv) == 0 {
			continue
		}
		if _, ok := processed[kv[0]]; ok {
			continue
		}

		// mark the key as existing.
		processed[kv[0]] = struct{}{}

		if len(returnTags) > 0 {
			returnTags += " "
		}
		returnTags += kv[0] + ":" + kv[1]
	}

	return returnTags
}

func TestRemoveDuplicate(t *testing.T) {
	type args struct {
		t string
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		{
			name: "Duplicate Test with 1",
			args: args{
				t: "json:\"name\"",
			},
			want: "json:\"name\"",
		},
		{
			name: "Duplicate Test with 2",
			args: args{
				t: "json:\"name\" json:\"name2\"",
			},
			want: "json:\"name\"",
		},
		{
			name: "Duplicate Test with 3",
			args: args{
				t: "json:\"name\" json:\"name2\" json:\"name3\"",
			},
			want: "json:\"name\"",
		},
		{
			name: "Duplicate Test with 3 and 1 unrelated",
			args: args{
				t: "json:\"name\" something:\"name2\" json:\"name3\"",
			},
			want: "json:\"name\" something:\"name2\"",
		},
		{
			name: "Duplicate Test with 3 and 2 unrelated",
			args: args{
				t: "something:\"name1\" json:\"name\" something:\"name2\" json:\"name3\"",
			},
			want: "something:\"name1\" json:\"name\"",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := removeDuplicateTags(tt.args.t); got != tt.want {
				t.Errorf("removeDuplicate() = %v, want %v", got, tt.want)
			}
		})
	}
}
```

File: plugin/modelgen/models.go [hunk 1]
```diff
@@ -415,9 +415,40 @@ func GoTagFieldHook(td *ast.Definition, fd *ast.FieldDefinition, f *Field) (*Fie
...
+				returnTags += " "
+			}
+			returnTags += k + ":" + v
+		}
+		return returnTags
+
+	}
+
+	return t
+}
```
Comment (StevenACoffman) [inline plugin/modelgen/models.go:450]:
```suggestion
// removeDuplicateTags removes duplicate tags
func removeDuplicateTags(t string) string {

	processed := make(map[string]bool)

	tt := strings.Split(t, " ")
	returnTags := ""

	for _, ti := range tt {
		kv := strings.Split(ti, ":")
		if len(kv) == 0 || processed[kv[0]] {
			continue
		}

		processed[kv[0]] = true

		if len(returnTags) > 0 {
			returnTags += " "
		}
		returnTags += kv[0] + ":" + kv[1]
	}

	return returnTags
}
```

File: plugin/modelgen/models_test.go [hunk 2]
```diff
@@ -350,3 +350,57 @@ func goBuild(t *testing.T, path string) error {
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := removeDuplicateTags(tt.args.t); got != tt.want {
+				t.Errorf("removeDuplicate() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
```
Comment (StevenACoffman) [inline plugin/modelgen/models_test.go:406]:
```suggestion
func TestRemoveDuplicate(t *testing.T) {
	type args struct {
		t string
	}
	tests := []struct {
		name string
		args args
		want string
	}{
		{
			name: "Duplicate Test with 1",
			args: args{
				t: "json:\"name\"",
			},
			want: "json:\"name\"",
		},
		{
			name: "Duplicate Test with 2",
			args: args{
				t: "json:\"name\" json:\"name2\"",
			},
			want: "json:\"name\"",
		},
		{
			name: "Duplicate Test with 3",
			args: args{
				t: "json:\"name\" json:\"name2\" json:\"name3\"",
			},
			want: "json:\"name\"",
		},
		{
			name: "Duplicate Test with 3 and 1 unrelated",
			args: args{
				t: "json:\"name\" something:\"name2\" json:\"name3\"",
			},
			want: "json:\"name\" something:\"name2\"",
		},
		{
			name: "Duplicate Test with 3 and 2 unrelated",
			args: args{
				t: "something:\"name1\" json:\"name\" something:\"name2\" json:\"name3\"",
			},
			want: "something:\"name1\" json:\"name\"",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := removeDuplicateTags(tt.args.t); got != tt.want {
				t.Errorf("removeDuplicate() = %v, want %v", got, tt.want)
			}
		})
	}
}
```

File: plugin/modelgen/models_test.go [hunk 3]
```diff
@@ -350,3 +350,57 @@ func goBuild(t *testing.T, path string) error {
...
+			name: "Duplicate Test with 3",
+			args: args{
+				t: "json:\"name\" json:\"name2\" json:\"name3\"",
+			},
+			want: "json:\"name3\"",
+		},
+		{
+			name: "Duplicate Test with 3 and 1 unrelated",
+			args: args{
+				t: "json:\"name\" something:\"name2\" json:\"name3\"",
```
Comment (StevenACoffman) [inline plugin/modelgen/models_test.go:387]:
```suggestion
				t: "json:\"name\" something:\"name2\" json:\"name3\"",
```

File: plugin/modelgen/models_test.go [hunk 4]
```diff
@@ -350,3 +350,57 @@ func goBuild(t *testing.T, path string) error {
...
+			},
+			want: "json:\"name3\" something:\"name2\"",
```
Comment (StevenACoffman) [inline plugin/modelgen/models_test.go:389]:
```suggestion
			want: "something:\"name2\" json:\"name3\"",
```
```suggestion
			want: "json:\"name3\" something:\"name2\"",
```

**Comment by StevenACoffman** [comment]
Hey, I hope it's ok, but I merged the tweaks I made in #2533 as they seemed a little easier to read. I would be happy to accept a PR if you feel like yours is better.

Thanks!

**Comment by valllabh** [comment]
Hey @StevenACoffman Thank you for taking a look at it.

Gqlgen can be used as just a model generator and in that case we may not have a full control over backend.
In such conditions `json:fieldName` can not be same as GoLang field name conventions.
Thats why my code was preferring to last value. It allows to override the values easily.
Other work around to override is to write own plugin and replace it during generation call, which is lengthy.

**Comment by StevenACoffman** [comment]
That makes sense! Hit me up with a PR if you feel like it should change to the behavior you originally proposed.

**Comment by valllabh** [comment]
Sure thing. Let me report issue and fork again

**Pull request closed without merging**